### PR TITLE
(Solid Start): Move common client types to core || remove tsup || createSession hook || support all cookies || rename `callbackUrl` to `redirectTo`

### LINKED ISSUE #6321: Cookies are not handled properly in the SolidStart integration
### Environment

```
  System:
    OS: macOS 13.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 173.56 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.12.1 - ~/.nvm/versions/node/v18.12.1/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v18.12.1/bin/yarn
    npm: 8.19.2 - ~/.nvm/versions/node/v18.12.1/bin/npm
  Browsers:
    Brave Browser: 108.1.46.144
    Firefox Developer Edition: 109.0
    Safari: 16.2
```

### Reproduction URL

https://github.com/nextauthjs/solid-start-auth-example

### Describe the issue

Due to a [SolidJS bug](https://github.com/solidjs/solid-start/issues/293) that prevents setting multiple `Set-Cookie` headers, the SolidStart integration [cannot be used](https://github.com/nextauthjs/next-auth/blob/3539a356010b952fefcad2a02ae3d60e2cba992d/packages/frameworks-solid-start/src/index.ts#L25-L43) with any provider that requires the `state` query parameter based on [RFC-6819](https://www.rfc-editor.org/rfc/rfc6819#section-4.4.1.8) to avoid CSRF attack against `redirect-uri`.

Although this is primarily a SolidJS bug, it would worth to _at least_ note this in the documentation, though it may would worth a joined effort to fix the issue as suggested in a [comment](https://github.com/solidjs/solid-start/issues/293#issuecomment-1283406131) on the tracked SolidJS issue.

Please ping me if more details or a better reproduction is needed. I did not create a separate repo for reproduction as the linked bug and code are both acknowledge that this issue exists.

### How to reproduce

1. Create a new SolidStart project
2. Add Auth.js integration
3. Add a provider with the `checks: ["state"]` option
4. Try to sign in

The following error should be raised:

```
CallbackRouteError: State was missing from the cookies.
    at Module.callback (********/node_modules/@auth/core/lib/routes/callback.js:243:23)
    at async AuthInternal (********/node_modules/@auth/core/lib/index.js:52:38)
    at async Auth (********/node_modules/@auth/core/index.js:95:30)
    at async ********/node_modules/@auth/solid-start/index.js:39:17
    at async Object.GET [as handler] (********/node_modules/@auth/solid-start/index.js:61:14)
    at async eval (/node_modules/solid-start/api/middleware.ts:21:16)
    at async eval (/node_modules/solid-start/entry-server/StartServer.tsx:45:12)
    at async devFetch (********/node_modules/solid-start/dev/server.js:45:12)
    at async startHandler (********/node_modules/solid-start/dev/server.js:112:20)
[auth][error][CallbackRouteError]: Read more at https://errors.authjs.dev#callbackrouteerror
```

### Expected behavior

The `next-auth.state` cookie should be set everywhere it is needed and the sign in flow should not raise any issues related to missing state cookie.

### PR DESCRIPTION
**Pull request by OrJDev**
> _NOTE_:
>
> - It's a good idea to open an issue first to discuss potential changes.
> - Please make sure that you are _NOT_ opening a PR to fix a potential security vulnerability. Instead, please follow the [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md) to disclose the issue to us confidentially.

## 🧢 Checklist

- [ ] Documentation
- [ ] Tests
- [ ] Ready to be merged

## 🎫 Affected issues

Please [scout and link issues](https://github.com/nextauthjs/next-auth/issues) that might be solved by this PR.

Fixes: INSERT_ISSUE_LINK_HERE

## 📌 Resources

- [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md)
- [Contributing guidelines](https://github.com/nextauthjs/.github/blob/main/CONTRIBUTING.md)
- [Code of conduct](https://github.com/nextauthjs/.github/blob/main/CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

closes https://github.com/nextauthjs/next-auth/issues/6321
add a nice `createSession` hook for solid-start, which will return the current session and will be updated correctly
remove the `next-auth` dependency from both solid-start and sveltekit

also move common client types to core so it will be

```ts
import type {
  LiteralUnion,
  SignInOptions,
  SignInAuthorizationParams,
  SignOutParams,
} from "@auth/core/types"
```

instead of 

```ts
import type {
  LiteralUnion,
  SignInOptions,
  SignInAuthorizationParams,
  SignOutParams,
} from "next-auth/react"
```

### CODE REVIEW

**Comment by OrJDev** [comment]
> I'm thinking about moving the whole client bit to the core now - curious what you think @balazsorban44 @OrJDev, as 
> 
> there's nothing relevant to each framework in those `lib/client.ts` files. 🤔 

Ye i also think so, i recently made another pr https://github.com/nextauthjs/next-auth/pull/6276 with a useSession hook and I couldn't use next auth (react) because I use Solid, and it seems to break things when using both solid and react so i moved the client types to /solid-start/src/types.ts

**Comment by ThangHuuVu** [comment]
@OrJDev would it make sense to do so in this PR too? 🙌 Since you're already moving some stuff around 👀

**Comment by OrJDev** [comment]
Ye sure, will update this pr probably today.

**Comment by OrJDev** [comment]
@ThangHuuVu i think its ready, what do you think? should i change anything?

Note: in the comments i say other pr - i meant the useSession hook - which i will need to update when this pr gets merged

File: apps/examples/solid-start/package.json [hunk 1]
```diff
@@ -18,7 +18,7 @@
   },
   "dependencies": {
     "@auth/core": "^0.1.4",
```
Comment (balazsorban44) [inline apps/examples/solid-start/package.json:20]: Let's set these to `latest`

File: apps/examples/solid-start/package.json [hunk 2]
```diff
@@ -18,7 +18,7 @@
...
-    "@solid-auth/next": "^0.0.19",
+    "@auth/solid-start": "^0.1.0",
```
Comment (balazsorban44) [inline apps/examples/solid-start/package.json:21]: Let's set these to `latest`

File: packages/core/package.json [hunk 3]
```diff
@@ -48,6 +48,10 @@
       "types": "./jwt.d.ts",
       "import": "./jwt.js"
     },
+    "./common": {
```
Comment (balazsorban44) [inline packages/core/package.json:51]: Let's not add another submodule, just use `@auth/core/types`

Comment (OrJDev) [reply packages/core/package.json:51]: not types, they are utils functions that help you create client functions

File: packages/core/src/common.ts [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+   * Nedded to reftech the session:
+   * res.ok && await __NEXTAUTH._getSession({ event: "storage" })
+   */
+  onFinish?: () => void | Promise<void>
+  /**
+   * prefix of the api route
+   */
+  prefix: string
+}
+export function createSignInClient(opts: CreateSignInClientOpts) {
```
Comment (balazsorban44) [inline packages/core/src/common.ts:46]: This should not be part of the core, as clients should be tailored for each framework. In almost no case this is the best way to do this in my opinion, as this uses client-side JS.

File: packages/core/src/common.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+      return
+    }
+    if (res.ok && opts.onSuccess) {
+      await opts.onSuccess()
+    }
+    return res
+  }
+}
+
+export function createSignOutClient(opts: CreateSignOutClientOpts) {
```
Comment (balazsorban44) [inline packages/core/src/common.ts:115]: Same as https://github.com/nextauthjs/next-auth/pull/6248/files#r1063979095

File: docs/docs/reference/04-solidstart/protected.md [hunk 6]
```diff
@@ -74,7 +74,7 @@ When using CSR, the `Protected` component will not work as expected and will cau
 
 ```tsx
 // entry-server.tsx
-import { Session } from "@auth/core";
+import { Session } from "@auth/core/types";
```
Comment (balazsorban44) [inline docs/docs/reference/04-solidstart/protected.md:77]:
```suggestion
import type { Session } from "@auth/core/types";
```

File: docs/docs/reference/04-solidstart/protected.md [hunk 7]
```diff
@@ -11,7 +11,7 @@ When using SSR, I recommend creating a `Protected` component that will trigger s
 
 ```tsx
 // components/Protected.tsx
-import { type Session } from "@auth/core";
+import { type Session } from "@auth/core/types";
```
Comment (balazsorban44) [inline docs/docs/reference/04-solidstart/protected.md:14]:
```suggestion
import type { Session } from "@auth/core/types";
```

File: packages/frameworks-solid-start/package.json [hunk 8]
```diff
@@ -15,14 +17,22 @@
     },
     "./client": {
       "types": "./client.d.ts",
-      "import": "./client.js"
+      "import": "./client.jsx"
+    },
+    "./types": {
+      "types": "./types.d.ts",
+      "import": "./types.js"
```
Comment (balazsorban44) [inline packages/frameworks-solid-start/package.json:24]:
```suggestion
      "types": "./types.d.ts"
```

File: packages/frameworks-solid-start/src/client.tsx [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+  useContext,
+} from "solid-js"
+import {
+  UseSessionOptions,
+  AuthClientConfig,
+  SessionProviderProps,
+} from "./types"
+import { now, objectIsSession, parseUrl } from "./utils"
+
+const __NEXTAUTH: AuthClientConfig = {
```
Comment (balazsorban44) [inline packages/frameworks-solid-start/src/client.tsx:28]: Should name things without `NEXT`

File: packages/frameworks-solid-start/src/client.tsx [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+
+  return value
+}
+
+/**
+ * Provider to wrap the app in to make session data available globally.
+ * Can also be used to throttle the number of requests to the endpoint
+ * `/api/auth/session`.
+ *
+ * [Documentation](https://next-auth.js.org/getting-started/client#sessionprovider)
```
Comment (balazsorban44) [inline packages/frameworks-solid-start/src/client.tsx:105]: Comments should be fixed, this is going to be a dead link. (Likely not need a link anymore as the docs page is generated from the same source)

File: packages/frameworks-solid-start/src/client.tsx [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+}
+
+/**
+ * Client-side method to initiate a signin flow
+ * or send the user to the signin page listing all possible providers.
+ * Automatically adds the CSRF token to the request.
+ *
+ * [Documentation](https://next-auth.js.org/getting-started/client#signin)
+ */
+export async function signIn<
```
Comment (balazsorban44) [inline packages/frameworks-solid-start/src/client.tsx:225]: I barely started looking into SolidStart. but why not Actions/Forms?

https://start.solidjs.com/core-concepts/actions

File: packages/core/src/types.ts [hunk 12]
```diff
@@ -473,3 +473,36 @@ export interface InternalOptions<TProviderType = ProviderType> {
...
+  | T
+  | (U & Record<never, never>)
+
+export interface SignInOptions extends Record<string, unknown> {
+  /**
+   * Specify to which URL the user will be redirected after signing in. Defaults to the page URL the sign-in is initiated from.
+   *
+   * [Documentation](https://next-auth.js.org/getting-started/client#specifying-a-callbackurl)
+   */
+  callbackUrl?: string
```
Comment (balazsorban44) [inline packages/core/src/types.ts:491]: Let's rename this to `redirectTo` (on the public side for now, we need to accommodate for this in core later).

We've been meaning to do this as it causes confusion:

https://github.com/nextauthjs/next-auth/issues/5305
https://github.com/nextauthjs/next-auth/issues/5308

File: packages/frameworks-solid-start/src/client.tsx (hunk 11 above)
Comment (OrJDev) [reply packages/frameworks-solid-start/src/client.tsx:225]: What is the point of form in this? We want people to be able to signIn from everywhere with every provider, a function is nice and can be used anywhere

File: packages/frameworks-solid-start/.gitignore [hunk 13]
```diff
@@ -2,6 +2,7 @@ node_modules
 dist
 **/*.d.ts
 **/*.js
+**/*.jsx
```
Comment (balazsorban44) [inline packages/frameworks-solid-start/.gitignore:5]: Are we shipping jsx? Will that work out of the box with SolidStart, or need an extra transpilation step?

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Added some comments

File: packages/frameworks-solid-start/.gitignore (hunk 13 above)
Comment (OrJDev) [reply packages/frameworks-solid-start/.gitignore:5]: Ye it works i already set the jsxImportSource to solid, and tried it. Seems to work fine

File: packages/frameworks-solid-start/src/client.tsx (hunk 11 above)
Comment (OrJDev) [reply packages/frameworks-solid-start/src/client.tsx:225]: Discussed this on twitter

File: packages/frameworks-solid-start/.gitignore (hunk 13 above)
Comment (OrJDev) [reply packages/frameworks-solid-start/.gitignore:6]: https://github.com/nextauthjs/next-auth/pull/6248/commits/5eea474c74037e1545f2f43fd10b4fa4a14d77ac

**Comment by gabor-boros** [comment]
@balazsorban44 @OrJDev is there anything I could help here to get it merged and unblock https://github.com/nextauthjs/next-auth/issues/6321? 😊

**Comment by OrJDev** [comment]
> @balazsorban44 @OrJDev is there anything I could help here to get it merged and unblock #6321? 😊

Thank you but it seems to be ready, we discussed on Twitter and he is pretty busy so he couldn't take a look at this pr, he will review and merge it when he is free.

File: packages/frameworks-solid-start/src/index.ts [hunk 14]
```diff
@@ -22,17 +22,21 @@ const actions: AuthAction[] = [
   "error",
 ]
 
-// currently multiple cookies are not supported, so we keep the next-auth.pkce.code_verifier cookie for now:
+// currently multiple cookies are not supported, so we keep the important ones
```
Comment (gabor-boros) [inline packages/frameworks-solid-start/src/index.ts:25]: @OrJDev It seems that solid-start got the multiple cookie support yesterday: https://github.com/solidjs/solid-start/pull/656

Comment (OrJDev) [reply packages/frameworks-solid-start/src/index.ts:25]: Awesome yes, i will update this pr today.

Comment (OrJDev) [reply packages/frameworks-solid-start/src/index.ts]: https://github.com/nextauthjs/next-auth/pull/6248/commits/20bc4325ad33580902469068aaa43262a1797c3e

**Comment by OrJDev** [comment]
focusing on solidstart
https://github.com/nextauthjs/next-auth/pull/7331
moved this pr to here

**Pull request closed without merging**